daemon/logger/etwlogs/etwlogs_windows.go

```diff
@@ -118,51 +117,43 @@ func unregisterETWProvider() {
 	mu.Lock()
 	defer mu.Unlock()
 	if refCount == 1 {
-		if callEventUnregister() {
-			refCount--
-			providerHandle = windows.InvalidHandle
+		if err := callEventUnregister(); err != nil {
+			// Not returning an error if EventUnregister fails, because etwLogs will continue to work
+			return
 		}
-		// Not returning an error if EventUnregister fails, because etwLogs will continue to work
+		refCount--
+		provider = nil
+		providerHandle = windows.InvalidHandle
 	} else {
 		refCount--
 	}
 }
 
-func callEventRegister() error {
-	// The provider's GUID is {a3693192-9ed6-46d2-a981-f8226c8363bd}
-	guid := windows.GUID{
-		Data1: 0xa3693192,
-		Data2: 0x9ed6,
-		Data3: 0x46d2,
-		Data4: [8]byte{0xa9, 0x81, 0xf8, 0x22, 0x6c, 0x83, 0x63, 0xbd},
-	}
-
-	ret, _, _ := procEventRegister.Call(uintptr(unsafe.Pointer(&guid)), 0, 0, uintptr(unsafe.Pointer(&providerHandle)))
-	if ret != win32CallSuccess {
-		errorMessage := fmt.Sprintf("Failed to register ETW provider. Error: %d", ret)
-		logrus.Error(errorMessage)
-		return errors.New(errorMessage)
+func callEventRegister() (*etw.Provider, error) {
+	providerID, _ := guid.FromString(providerGUID)
+	p, err := etw.NewProviderWithOptions("", etw.WithID(providerID))
+	if err != nil {
+		logrus.WithError(err).Error("Failed to register ETW provider")
+		return nil, fmt.Errorf("failed to register ETW provider: %v", err)
 	}
-	return nil
+	return p, nil
 }
 
+// TODO(thaJeztah): use provider.WriteEvent()
 func callEventWriteString(message string) error {
```

Comment:
Need to look into this one; so we currently take a _structured_ message, and convert it to a string, then write that string to etwlogs (perhaps etwlogs internally converts it back to a structured message based on a specific format??); https://github.com/moby/moby/blob/a189651b4c772881ba87eec181312cd8bc20ec2e/daemon/logger/etwlogs/etwlogs_windows.go#L93-L101


Whereas ` provider.WriteEvent()`  provides functionality to write a _structured_ event; https://github.com/moby/moby/blob/a189651b4c772881ba87eec181312cd8bc20ec2e/vendor/github.com/Microsoft/go-winio/pkg/etw/provider.go#L208-L211

So, I think we should switch to use that, but I'm not sure if that would change things for the enduser (does the current `EventWriteString` convert it back to a structured message? If so, changing will probably not change anything to the user.

Alternatively, we could contribute a `WriteString` method in go-winio (if we can't change)

Reply:
We can have a look at the logrus hook how to do the conversion (if we can); https://github.com/moby/moby/blob/a189651b4c772881ba87eec181312cd8bc20ec2e/vendor/github.com/Microsoft/go-winio/pkg/etwlogrus/hook.go#L75-L96

Reply:
I guess what we're missing is a `name` (first argument of `provider.WriteEvent()`); for the Logrus hook, that's hardcoded to `LogrusEntry`, which feels a bit odd to use (but I don't know how/where these show up).